clang/test/CIR/CodeGen/stmt-expr.c

```diff
@@ -0,0 +1,37 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+// Yields void.
+void test1() { ({ }); }
+// CHECK: @test1
+//     CHECK: cir.scope {
+// CHECK-NOT:   cir.yield
+//     CHECK: }
+
+// Yields an l-value.
+void test2(int x) { ({ x;}); }
+// CHECK: @test2
+// CHECK: %{{.+}} = cir.scope {
+// CHECK:   %[[#V6:]] = cir.alloca !s32i, cir.ptr <!s32i>, ["tmp"]
+// CHECK:   %[[#V7:]] = cir.load %{{.+}} : cir.ptr <!s32i>, !s32i
+// CHECK:   cir.store %[[#V7]], %[[#V6]] : !s32i, cir.ptr <!s32i>
+// CHECK:   cir.yield %[[#V6]] : !cir.ptr<!s32i>
+// CHECK: }
+
+// Yields an aggregate.
+struct S { int x; };
+void test4() { ({ struct S s = {1}; s; }); }
+// CHECK: @test4
+// CHECK: %[[#RET:]] = cir.alloca !ty_22S22, cir.ptr <!ty_22S22>
+// CHECK: cir.scope {
+// CHECK:   %[[#VAR:]] = cir.alloca !ty_22S22, cir.ptr <!ty_22S22>
+//          [...]
+// CHECK:   cir.copy %[[#VAR]] to %[[#RET]] : !cir.ptr<!ty_22S22>
+// CHECK: }
+
```

Comment:
I'd also like to see the codegen for two examples from https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html:

1.
```
A a;
({a;}).Foo ()
```

and 

2.
```
#define macro(a)  ({__typeof__(a) b = (a); b + 3; })
template<typename T> T function(T a) { T b = a; return b + 3; }

void foo ()
{
  macro (X ());
  function (X ());
}
```

It's really important for later lifetime-checking we get the construction right here.

Reply:
The first example is generated as expected by the documentation: the local `a` is allocated and constructed, then copy-constructed into a temporary, which is used to call `Foo`.

<details>
  <summary>Codegen of the First Example</summary>

### Input Code

```c++
class A {
public:
  A() {}
  void Foo() {}
};

void test1() {
  A a;
  ({a;}).Foo();
}
```  

### CIR Code

```mlir
!u8i = !cir.int<u, 8>
!ty_22A22 = !cir.struct<class "A" {!cir.int<u, 8>}>
module {
  cir.func linkonce_odr @_ZN1AC2Ev(%arg0: !cir.ptr<!ty_22A22>) extra( {inline = #cir.inline<no>, optnone = #cir.optnone} ) {
    %0 = cir.alloca !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>, ["this", init] {alignment = 8 : i64}
    cir.store %arg0, %0 : !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>
    %1 = cir.load %0 : cir.ptr <!cir.ptr<!ty_22A22>>, !cir.ptr<!ty_22A22>
    cir.return
  }
  cir.func linkonce_odr @_ZN1AC2ERKS_(%arg0: !cir.ptr<!ty_22A22>, %arg1: !cir.ptr<!ty_22A22>) extra( {inline = #cir.inline<no>, optnone = #cir.optnone} ) {
    %0 = cir.alloca !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>, ["this", init] {alignment = 8 : i64}
    %1 = cir.alloca !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>, ["", init] {alignment = 8 : i64}
    cir.store %arg0, %0 : !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>
    cir.store %arg1, %1 : !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>
    %2 = cir.load %0 : cir.ptr <!cir.ptr<!ty_22A22>>, !cir.ptr<!ty_22A22>
    cir.return
  }
  cir.func linkonce_odr @_ZN1A3FooEv(%arg0: !cir.ptr<!ty_22A22>) extra( {inline = #cir.inline<no>, optnone = #cir.optnone} ) {
    %0 = cir.alloca !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>, ["this", init] {alignment = 8 : i64}
    cir.store %arg0, %0 : !cir.ptr<!ty_22A22>, cir.ptr <!cir.ptr<!ty_22A22>>
    %1 = cir.load %0 : cir.ptr <!cir.ptr<!ty_22A22>>, !cir.ptr<!ty_22A22>
    cir.return
  }
  cir.func @_Z5test1v() extra( {inline = #cir.inline<no>, optnone = #cir.optnone} ) {
    %0 = cir.alloca !ty_22A22, cir.ptr <!ty_22A22>, ["a", init] {alignment = 1 : i64}
    cir.call @_ZN1AC2Ev(%0) : (!cir.ptr<!ty_22A22>) -> ()
    cir.scope {
      %1 = cir.alloca !ty_22A22, cir.ptr <!ty_22A22>, ["ref.tmp0"] {alignment = 1 : i64}
      cir.scope {
        cir.call @_ZN1AC2ERKS_(%1, %0) : (!cir.ptr<!ty_22A22>, !cir.ptr<!ty_22A22>) -> ()
      }
      cir.call @_ZN1A3FooEv(%1) : (!cir.ptr<!ty_22A22>) -> ()
    }
    cir.return
  }
}
```

</details>

I can't codegen the second example since destructors are currently broken. Still, I would guess it is not working as expected by the documentation (e.g. destructor of the temporary called right after the expression, not the end of the statement).

Reply:
Thanks for the explanation, looks great. We can look into option 2 when dtors improve a bit, I fixed lots of issues last week and there are many more.